### PR TITLE
[SPARK-5710][SQL] Combines two adjacent Cast expressions into one

### PR DESCRIPTION
**Pull request by guowei2**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -602,6 +603,16 @@ object CombineLimits extends Rule[LogicalPlan] {
 }
 
 /**
+ * Combines two adjacent [[Cast]] expressions into one
+ */
+object CombineCasts extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+    case cc @ Cast(c2 @ Cast(c1, dataType1), dataType2) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: `cc` and `c2` is not used. You could just write `case Cast(Cast(c1, dataType1), dataType2)`

**Comment by chenghao-intel** [comment]
This is probably work for some cases, not for all of the cases. For example:
Int => Data => String is not equals to Int => String

**Comment by guowei2** [comment]
@chenghao-intel  
can we cast int or long to Date?

**Comment by guowei2** [comment]
@chenghao-intel  i think i should restrict it in numeric.
```
object CombineCasts extends Rule[LogicalPlan] {
  def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
    case Cast(Cast(c1, dataType1: NumericType), dataType2: NumericType) =>
      Cast(c1, dataType2)
  }
}
```

**Comment by rxin** [comment]
Numeric casts can lose precision, and as a result, I don't think the cast works. 

e.g.

long => int => long yields different results from long => long

**Comment by rxin** [comment]
Did you see a query that needs this optimization?

**Comment by guowei2** [comment]
@rxin 
For example. 
create table test(a decimal(3,1));
explain select * from test where a*2-1>1;
== Physical Plan ==
Filter (CAST(CAST((CAST(CAST((CAST(a#5, DecimalType()) * 2), DecimalType(21,1)), DecimalType()) - 1), DecimalType(22,1)), DecimalType()) > 1)
HiveTableScan a#5, (MetastoreRelation default, test, None), None

**Comment by srowen** [comment]
@guowei2 can this change be limited to merging adjacent casts that are definitely redundant? it sounds like this change introduces bugs as-is, but there may still be some opportunities for optimization. It'd be good to clarify whether this is proceeding or not.

**Comment by marmbrus** [comment]
Agreed.  If there are concrete proposals for eliminating redundant casts then we should discuss on JIRA.  However as is this could change the answer and thus is an invalid optimization.  So, we should close this issue.

**Pull request closed without merging**